src/Compiler/Opts/ConstantFold.idr

```diff
@@ -0,0 +1,153 @@
+        replace {p=CExp} (sym $ appendAssociative args outer inner) $
+        inlineLetEnv (mkSizeOf args + out) val $
+        replace {p=CExp} (appendAssociative args outer ([x] ++ inner)) w
+
+    inlineLetConstAlt : SizeOf outer -> CExp inner -> CConstAlt (outer ++ [x] ++ inner) -> CConstAlt (outer ++ inner)
+    inlineLetConstAlt out val (MkConstAlt c e) = MkConstAlt c $ inlineLetEnv out val e
+
+  export
+  inlineLet : CExp vars -> CExp (x :: vars) -> CExp vars
+  inlineLet val exp = inlineLetEnv zero val exp
+
+findConstAlt : Constant -> List (CConstAlt vars) -> Maybe (CExp vars) -> Maybe (CExp vars)
+findConstAlt c [] def = def
+findConstAlt c (MkConstAlt c' exp :: alts) def = if c == c'
+    then Just exp
+    else findConstAlt c alts def
+
+foldableOp : PrimFn ar -> Bool
+foldableOp BelieveMe = False
+foldableOp (Cast from to) = fromMaybe False [| intKind from `smaller` intKind to |]
+  where
+    smaller : IntKind -> IntKind -> Bool
+    smaller (Signed x) (Signed y) = x <= y
+    smaller (Unsigned x) (Unsigned y) = x <= y
+    smaller (Signed x) (Unsigned y) = x < P y
+    smaller (Unsigned x) (Signed y) = P x < y
+foldableOp _ = True
+
+-- constant folding of primitive functions
+-- if a primitive function is applied to only constant
+-- then replace with the result
+-- if there's only 1 constant argument to a commutative function
+-- move the constant to the right. This simplifies Compiler.Opts.Identity
+constFold : {vars : _} -> CExp vars -> CExp vars
+constFold e@(CLocal _ _) = e
+constFold e@(CRef fc x) = e
+constFold (CLam fc x y) = CLam fc x $ constFold y
+constFold (CLet fc x inlineOK y z) =
+    let val = constFold y
+     in case val of
+        CPrimVal _ _ => if inlineOK
+            then constFold $ inlineLet val z
+            else CLet fc x inlineOK val (constFold z)
+        _ => CLet fc x inlineOK val (constFold z)
+constFold (CApp fc x xs) = CApp fc (constFold x) (constFold <$> xs)
+constFold (CCon fc x y tag xs) = CCon fc x y tag $ constFold <$> xs
+constFold (COp {arity} fc fn xs) =
+    let xs' : Vect arity (CExp vars)
+        xs' = map constFold xs
+        e : CExp vars
+        e = constRight fc fn xs'
+     in if foldableOp fn
+        then case the (Maybe (Vect arity (NF vars))) $ Prelude.traverse toNF xs' of
+                Nothing => e
+                Just nfs => case getOp fn nfs of
+                    Just nf => fromMaybe e $ fromNF nf
+                    Nothing => e
+        else e
+  where
+    toNF : CExp vars -> Maybe (NF vars)
+    toNF (CPrimVal fc (I _)) = Nothing -- don't fold `Int` because it has varying widths
+    toNF (CPrimVal fc c) = Just $ NPrimVal fc c
+    toNF _ = Nothing
+
+    fromNF : NF vars -> Maybe (CExp vars)
+    fromNF (NPrimVal fc c) = Just $ CPrimVal fc c
+    fromNF _ = Nothing
+
+    constRight : {ar : _} -> FC -> PrimFn ar -> Vect ar (CExp vars) -> CExp vars
+    constRight fc (Add ty) [x@(CPrimVal _ _), y] = COp fc (Add ty) [constFold y, x]
```

Comment:
Same comment here: I don't think we can guarantee that addition on floats is commutative.